lib/backend/etcdv3/etcdv3.go

```diff
@@ -0,0 +1,465 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package etcdv3
+
+import (
+	"context"
+	goerrors "errors"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/coreos/etcd/mvcc/mvccpb"
+	"github.com/coreos/etcd/pkg/transport"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/projectcalico/libcalico-go/lib/apiconfig"
+	"github.com/projectcalico/libcalico-go/lib/backend/api"
+	"github.com/projectcalico/libcalico-go/lib/backend/model"
+	"github.com/projectcalico/libcalico-go/lib/errors"
+)
+
+var (
+	clientTimeout = 30 * time.Second
+)
+
+type EtcdV3Client struct {
+	etcdClient *clientv3.Client
+}
+
+func NewEtcdV3Client(config *apiconfig.EtcdConfig) (api.Client, error) {
+	// Determine the location from the authority or the endpoints.  The endpoints
+	// takes precedence if both are specified.
+	etcdLocation := []string{}
+	if config.EtcdEndpoints != "" {
+		etcdLocation = strings.Split(config.EtcdEndpoints, ",")
+	}
+
+	if len(etcdLocation) == 0 {
+		log.Info("No etcd endpoints specified in etcdv3 API config")
+		return nil, goerrors.New("no etcd endpoints specified")
+	}
+
+	// Create the etcd client
+	tlsInfo := &transport.TLSInfo{
+		CAFile:   config.EtcdCACertFile,
+		CertFile: config.EtcdCertFile,
+		KeyFile:  config.EtcdKeyFile,
+	}
+	tls, _ := tlsInfo.ClientConfig()
+
+	cfg := clientv3.Config{
+		Endpoints:   etcdLocation,
+		TLS:         tls,
+		DialTimeout: clientTimeout,
+	}
+
+	// Plumb through the username and password if both are configured.
+	if config.EtcdUsername != "" && config.EtcdPassword != "" {
+		cfg.Username = config.EtcdUsername
+		cfg.Password = config.EtcdPassword
+	}
+
+	client, err := clientv3.New(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	return &EtcdV3Client{etcdClient: client}, nil
+}
+
+// Create an entry in the datastore.  If the entry already exists, this will return
+// an ErrorResourceAlreadyExists error and the current entry.
+func (c *EtcdV3Client) Create(d *model.KVPair) (*model.KVPair, error) {
+	logCxt := log.WithFields(log.Fields{"model-key": d.Key, "value": d.Value, "ttl": d.TTL, "rev": d.Revision})
+	logCxt.Debug("Processing Create request")
+	key, value, err := getKeyValueStrings(d)
+	if err != nil {
+		return nil, err
+	}
+	logCxt = logCxt.WithField("etcdv3-key", key)
+
+	putOpts, err := c.getTTLOption(d)
+	if err != nil {
+		return nil, err
+	}
+
+	// Checking for 0 version of the key, which means it doesn't exists yet,
+	// and if it does, get the current value.
+	logCxt.Debug("Performing etcdv3 transaction for Create request")
+	txnResp, err := c.etcdClient.Txn(context.Background()).If(
```

Comment:
should we have some sort of client timeout for the context (different from the `DialTimeout`)? 
```go
ctx, cancel := context.WithTimeout(context.Background(), timeout)
txnResp, err := c.etcdClient.Txn(ctx).If(
...
```

Reply:
yes we should.  can you ignore that for now though - I was planning on handling the watch/list timeouts as part of another task.

Reply:
ok as long as you're keeping track of those :)